src/nvim/api/autocmd.c

```diff
@@ -0,0 +1,669 @@
+      api_set_error(err, kErrorTypeException, "Failed to set augroup");
+      return -1;
+    }
+
+    if (clear_autocmds) {
+      FOR_ALL_AUEVENTS(event) {
+        aupat_del_for_event_and_group(event, augroup);
+      }
+    }
+  });
+
+  return augroup;
+}
+
+/// NOTE: behavior differs from augroup-delete.
+/// When deleting an augroup, autocmds contained by this augroup will also be deleted and cleared.
+/// This augroup will no longer exist
+void nvim_del_augroup_by_id(Integer id)
+  FUNC_API_SINCE(9)
+{
+  char *name = augroup_name((int)id);
+  augroup_del(name, false);
+}
+
+/// NOTE: behavior differs from augroup-delete.
+/// When deleting an augroup, autocmds contained by this augroup will also be deleted and cleared.
+/// This augroup will no longer exist
+void nvim_del_augroup_by_name(String name)
+  FUNC_API_SINCE(9)
+{
+  augroup_del(name.data, false);
+}
+
+/// -- @param {string} group - autocmd group name
+/// -- @param {number} buffer - buffer number
+/// -- @param {string | table} event - event or events to match against
+/// -- @param {string | table} pattern - optional, defaults to "*".
+/// vim.api.nvim_do_autcmd({ group, buffer, pattern, event, modeline })
+void nvim_do_autocmd(Dict(do_autocmd) *opts, Error *err)
+  FUNC_API_SINCE(9)
+{
+  int au_group = AUGROUP_ALL;
+  bool modeline = true;
+
+  buf_T *buf = curbuf;
+  bool set_buf = false;
+
+  char_u *pattern = NULL;
+  bool set_pattern = false;
+
+  Array event_array = ARRAY_DICT_INIT;
+
+  if (opts->group.type != kObjectTypeNil) {
+    if (opts->group.type != kObjectTypeString) {
+      api_set_error(err, kErrorTypeValidation, "'group' must be a string");
+      goto cleanup;
+    }
+
+    au_group = augroup_find(opts->group.data.string.data);
+
+    if (au_group == AUGROUP_ERROR) {
+      api_set_error(err,
+                    kErrorTypeException,
+                    "invalid augroup: %s", opts->group.data.string.data);
+
+      goto cleanup;
+    }
+  }
+
+  if (opts->buffer.type != kObjectTypeNil) {
+    Object buf_obj = opts->buffer;
+    if (buf_obj.type != kObjectTypeInteger && buf_obj.type != kObjectTypeBuffer) {
+      api_set_error(err, kErrorTypeException, "invalid buffer: %d", buf_obj.type);
+      goto cleanup;
+    }
+
+    buf = find_buffer_by_handle((Buffer)buf_obj.data.integer, err);
+    set_buf = true;
+
+    if (ERROR_SET(err)) {
+      goto cleanup;
+    }
+  }
+
+  if (opts->pattern.type != kObjectTypeNil) {
+    if (opts->pattern.type != kObjectTypeString) {
+      api_set_error(err, kErrorTypeValidation, "'pattern' must be a string");
+      goto cleanup;
+    }
+
+    pattern = vim_strsave((char_u *)opts->pattern.data.string.data);
+    set_pattern = true;
+  }
+
+  if (opts->event.type != kObjectTypeNil) {
+    UNPACK_STRING_OR_ARRAY(event_array, "event", (&opts->event), cleanup)
+  }
+
+  if (opts->modeline.type != kObjectTypeNil) {
+    modeline = api_object_to_bool(opts->modeline, "modeline", true, err);
```

Comment:
no need to NIL check around `api_object_to_bool`. The default value used for nil is supplied.